runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/internal/StateMultiplexingGroupByKey.java

```diff
@@ -0,0 +1,316 @@
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.transforms.windowing.AfterWatermark.AfterWatermarkEarlyAndLate;
+import org.apache.beam.sdk.transforms.windowing.AfterWatermark.FromEndOfWindow;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.transforms.windowing.DefaultTrigger;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindows;
+import org.apache.beam.sdk.transforms.windowing.KeyedWindow;
+import org.apache.beam.sdk.transforms.windowing.KeyedWindow.KeyedWindowFn;
+import org.apache.beam.sdk.transforms.windowing.Never.NeverTrigger;
+import org.apache.beam.sdk.transforms.windowing.PaneInfo;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.transforms.windowing.WindowFn;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollection.IsBounded;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.apache.beam.sdk.values.WindowingStrategy;
+import org.apache.beam.vendor.grpc.v1p60p1.com.google.common.base.Preconditions;
+import org.apache.beam.vendor.grpc.v1p60p1.com.google.protobuf.ByteString;
+import org.apache.beam.vendor.grpc.v1p60p1.com.google.protobuf.ByteString.Output;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.Iterables;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * A GroupByKey implementation that multiplexes many small user keys over a fixed set of sharding
+ * keys for reducing per key overhead.
+ */
+public class StateMultiplexingGroupByKey<K, V>
+    extends PTransform<PCollection<KV<K, V>>, PCollection<KV<K, Iterable<V>>>> {
+
+  /*
+   * Keys larger than this threshold will not be multiplexed.
+   */
+  private static final int SMALL_KEY_BYTES_THRESHOLD = 4096;
+  private final boolean fewKeys;
+  private final int numShardingKeys;
+
+  private StateMultiplexingGroupByKey(boolean fewKeys) {
+    // :TODO plumb fewKeys to DataflowGroupByKey
+    this.fewKeys = fewKeys;
+    // :TODO Make this configurable
+    this.numShardingKeys = 32 << 10;
+  }
+
+  /**
+   * Returns a {@code GroupByKey<K, V>} {@code PTransform}.
+   *
+   * @param <K> the type of the keys of the input and output {@code PCollection}s
+   * @param <V> the type of the values of the input {@code PCollection} and the elements of the
+   *     {@code Iterable}s in the output {@code PCollection}
+   */
+  public static <K, V> StateMultiplexingGroupByKey<K, V> create(boolean fewKeys) {
+    return new StateMultiplexingGroupByKey<>(fewKeys);
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+
+  public static void applicableTo(PCollection<?> input) {
```

Comment:
is this copied from somewhere else? can we refactor to share it so we don't have implementations drift?

Reply:
done